# Variance Analysis

Variance Analysis is used when trying to find out whether there is a statistically significant difference between the means of two or more groups.

## Prerequisites of Variance Analysis

 * 1-)Samples are independent of each other
 * 2-)Samples should have normal distribution
 * 3-) The distribution of variance of thesamples should be similar (variance homogeneity)


**Business Application:** Determining of  homepage content of website
 
**Target:** 

* In order to increase the time that users spend on the website, it is desired to determine the contents on the homepage of the website.

**Details:**

* It is known that users who spend a long time on the website click more on advertisements and their loyalty to the brand increases.

* For this purpose, similar contents are prepared with different pictures or in different formats. These created contents are shown to different test groups.

* Contents ara presented in three ways:

* **Group A**: With the Natural way
* **Group B**: With leading  way
* **Group C**: With intersting way




    

**H0:** There **is no** statistically significant difference in the average time spent on the website of the groups. 
    
    
**H1:**  there **is**  statistically significant difference in the average time spent on the website of the groups. 

In [1]:
import pandas as pd
A = pd.DataFrame([28,33,30,29,28,29,27,31,30,32,28,33,25,29,27,31,31,30,31,34,30,32,31,34,28,32,31,28,33,29])# the  time  spent on the website for group A

B = pd.DataFrame([31,32,30,30,33,32,34,27,36,30,31,30,38,29,30,34,34,31,35,35,33,30,28,29,26,37,31,28,34,33])# the  time spent on the website for group B


C = pd.DataFrame([40,33,38,41,42,43,38,35,39,39,36,34,35,40,38,36,39,36,33,35,38,35,40,40,39,38,38,43,40,42])# the time spent on the website for group C


dfs = [A, B, C]

ABC = pd.concat(dfs, axis = 1)
ABC.columns = ["GROUP_A","GROUP_B","GROUP_C"]
ABC.head()

GROUP_A  GROUP_B  GROUP_C
0       28       31       40
1       33       32       33
2       30       30       38
3       29       30       41
4       28       33       42

## Checking the  prerequisites for appplying this test

1-)Samples should have normal distribution

2-) The distribution of variance of the two samples should be similar (variance homogeneity)

In [3]:
from scipy.stats import shapiro

### 1-)Normal distribution

**Shapiro-Wilks Test**

**HO:** There **is no** statistically significant difference between the sample distribution and the theoretical normal distribution.

**H1:** There **is** a statistically significant difference between the sample distribution and the theoretical normal distribution. 

In [5]:
shapiro(ABC["GROUP_A"])

ShapiroResult(statistic=0.969743549823761, pvalue=0.5321813225746155)

**pvalue=0.53218132**

* Since the sample p-value is not less than 0.05, we cannot reject the H0 hypothesis.

In [6]:
shapiro(ABC["GROUP_B"])

ShapiroResult(statistic=0.9789854884147644, pvalue=0.7979801297187805)

**pvalue=0.7979801**

* Since the sample p-value is not less than 0.05, we cannot reject the H0 hypothesis.

In [7]:
shapiro(ABC["GROUP_C"])

ShapiroResult(statistic=0.9579201340675354, pvalue=0.273820161819458)

**pvalue=0.2738201**

* Since the sample p-value is not less than 0.05, we cannot reject the H0 hypothesis.

### 2-)Variance homogeneity

**H0:** The variances of samples **are** homogeneous


**H1:** The variances of  samples **are not** homogeneous

In [13]:
import scipy.stats as stats
stats.levene(ABC["GROUP_A"], ABC["GROUP_B"],ABC["GROUP_C"])

LeveneResult(statistic=1.0267403645055275, pvalue=0.36247110117417064)

* Since the sample p-value is not less than 0.05, we cannot reject the H0 hypothesis.


* Thus the variances of both samples are homogeneous

## Hypothesis Test

**H0:** There **is no** statistically significant difference in the average time spent on the website of the groups. 
    
    
**H1:**  there **is**  statistically significant difference in the average time spent on the website of the groups. 

In [15]:
from scipy.stats import f_oneway

In [16]:
f_oneway(ABC["GROUP_A"], ABC["GROUP_B"],ABC["GROUP_C"])

F_onewayResult(statistic=74.69278140730431, pvalue=1.3079050746811477e-19)

In [18]:
print("p value:" +'{:.5f}'.format(f_oneway(ABC["GROUP_A"], ABC["GROUP_B"],ABC["GROUP_C"])[1]))

p value:0.00000


**p value:0.00000**
    
* H0 hypothesis is rejected because the p-value is less than 0.05.


* thus  there **is**  statistically significant difference in the average time spent on the website of the groups.

In [19]:
ABC.describe().T

count       mean       std   min    25%   50%    75%   max
GROUP_A   30.0  30.133333  2.224214  25.0  28.25  30.0  31.75  34.0
GROUP_B   30.0  31.700000  2.937862  26.0  30.00  31.0  34.00  38.0
GROUP_C   30.0  38.100000  2.808239  33.0  36.00  38.0  40.00  43.0

# 2-)Non-Parametric Hypothesis Test

If the normal distribution and variance homogeneity as prerequisites **are not satisfied**, we should apply Non-Parametric Hypothesis Test on samples.

In [20]:
from scipy.stats import kruskal

In [21]:
kruskal(ABC["GROUP_A"], ABC["GROUP_B"],ABC["GROUP_C"])

KruskalResult(statistic=54.19819735523783, pvalue=1.7022015426175926e-12)